### DISCUSSION TITLE
Import and export TS types and interfaces in Next.js app don't needed anymore?

### DISCUSSION THREAD

**Discussion by AngelinaPlax** [vercel/next.js#50684] [Help] [answered]
### Summary

Hello!
I study Next.js and create an app based on Next.js and TypeScript.
I don't add import and export to my TS interfaces and types, that are stored in another folder, and it works!
Can I don't use import for TS interfaces and types?
(But if I add export to my interface, error will appeared in file, where I use this interface)

If I have this code in types.ts file:
```
interface IAboutIdProps {
  params: {
    id: string;
  };
}
```
and this code in my page.tsx file:
```
import { shops } from "@/app/constants/constants";

export default function AboutId({ params: { id } }: IAboutIdProps) {
  const shopInfo = shops.find((shopInfo) => shopInfo.id === id);
  return (
    <h3>
      Something about {shopInfo?.name || "This shop does not exist in DB"}
    </h3>
  );
}
```
It will work.
But when I do this way:
```
export interface IAboutIdProps {
  params: {
    id: string;
  };
}
```
my page file will have issue 'Cannot find name 'IAboutIdProps''

Could you please explain me why it works like that?

**Comment by devjiwonchoi** [comment]
Did you import the `types` file to the page?

```js
import { shops } from "@/app/constants/constants";
import { IAboutIdProps } from "@/app/types";  // location of the `types` file.

export default function AboutId({ params: { id } }: IAboutIdProps) {
  const shopInfo = shops.find((shopInfo) => shopInfo.id === id);
  return (
    <h3>
      Something about {shopInfo?.name || "This shop does not exist in DB"}
    </h3>
  );
}
```

**Reply by AngelinaPlax** [reply]
I did it, of course.
Question about why my application works even if I do not add imports and exports for my TS types

**Comment by icyJoseph** [comment, marked as answer]
This is an obscure TypeScript _feature_. If a file has no import/export, then it is not a module, so it becomes global.

For example:

```ts
// import type {GetServerSideProps} from 'next'

interface Foo {
  bar(): void;
}

type Baz = {
  fizz(): void;
};

```

As long as the top line is commented, Foo and Baz will be visible from other files. The moment you uncomment that, this becomes a module and both Foo, and Baz, become private instances of the module. Likewise, the moment this file does an `export`, it turns into a module too.

This sort of reliance on global, or magically available, types, is anti-pattern that'll bring lots of chaos to your codebase. Avoid it. 

## --isolatedModules

> https://www.typescriptlang.org/tsconfig#isolatedModules

What is throwing me off here, is that `isolatedModules` should forbid this, but it is not 😅 😮 - Perhaps I am missing something else, but it should be like, parsers, should be able, by looking at a single file, to get all type information they need to compile that file, and with "magically available" types that's not possible.

This is what I'd expect:

```console
'foo.ts' cannot be compiled under '--isolatedModules' because it is considered a global script file. Add an import, export, or an empty 'export {}' statement to make it a module.ts(1208)
```

For a `foo.ts` file:

```ts
function foo(){}
```

## typescript@4.5 (verified in 4.9.5 too)

It might be a TS issue... now that I've downgraded from 5.x.x to 4.x.x, I see the error I'd expect:

![Screenshot 2023-06-02 at 14 33 25](https://github.com/vercel/next.js/assets/21013447/7219adef-5e02-4da8-a009-5ea520e9326a)

I'll go through the TS release logs and see if there's something mentioning this change in behaviour.

For completeness sake here's the entire file from the screenshot:

```ts
// import type {GetServerSideProps} from 'next'

interface Foo {
  bar(): void;
}

type Baz = {
  fizz(): void;
};

function foo() {}
```

**Reply by icyJoseph** [reply]
Alright, managed to test up to `typescript@4.9.5` and it works up till then. Tested with `typescript@5.0.2` and it is starts to show this behavior.

**Reply by icyJoseph** [reply]
Alright @AngelinaPlax according to this issue: https://github.com/microsoft/TypeScript/issues/53683, it is working as expected 😅 I wonder if there's a combination of flags to still trigger a compilation error.

**Reply by icyJoseph** [reply]
One thing you can do now is add:

```json
"moduleDetection": "force",
```

To your tsconfig file, and that'll forbid the behavior you observed @AngelinaPlax

**Reply by AngelinaPlax** [reply]
Many thanks!
